### PR TITLE
Always try to resolve conversion for Any type

### LINKED ISSUE #5898: Any.== slowed down by Panic.catch 2x times
I have just been verifying the effect of #5845 on performance of [sieve.enso](https://github.com/jtulach/sieve/blob/5ff0f8881aed3d0b772c532e22665361ad8f0854/enso/sieve.enso) and via _VisualVM Polyglot Sampler_ I can see that an enormous time is being spent on [line 31 comparing == 0](https://github.com/jtulach/sieve/blob/5ff0f8881aed3d0b772c532e22665361ad8f0854/enso/sieve.enso#L31) and subsequently catching panics. Single round of the benchmark takes:
```
Hundred thousand prime numbers in 895 ms. Last one is 1299709
```
By changing the Enso standard library to:
```diff
diff --git distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso
index 565b64adcb..7181423508 100644
--- distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso
+++ distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso
@@ -108,8 +108,8 @@ type Any
     == self that =
         # If there is No_Such_Conversion, then `self` and `that` are probably
         # host or polyglot values, so we just compare them with the default comparator.
-        eq_self = Panic.catch No_Such_Conversion (Comparable.from self) _-> Default_Comparator
-        eq_that = Panic.catch No_Such_Conversion (Comparable.from that) _-> Default_Comparator
+        eq_self = Comparable.from self
+        eq_that = Comparable.from that
         similar_type = Meta.is_same_object eq_self eq_that
         case similar_type of
             False -> False
```
I am able to get to better numbers. Single round of the benchmark goes down to:
```
Hundred thousand prime numbers in 403 ms. Last one is 1299709
```
e.g. the benchmark gets twice as fast by eliminating `Panic.catch` call.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Fixes #5898 by removing `Catch.panic` and speeding the `sieve.enso` benchmark from 1058 ms to 514 ms. Should there be no dedicated conversion, let's use one defined on `Any` type - e.g. defining a conversion `from(Any)` makes such a conversion is always available.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

**Review by radeusgd — COMMENTED** [review]
Looks good.

Maybe worth adding a few test cases for it?

I'd imagine at least 3:
1. Have a conversion both on type `Foo` and `Any`: verify that for an instance of type `Foo` the more specific one (`Foo`, not `Any`) is selected.
2. Have a conversion on `Any` but not `Foo` - check that `Any` is selected - essentially what this PR does and is implicitly tested by the `==` not breaking, but I think it's good to have an explicit test for it.
3. Have no conversion defined on neither `Foo` nor `Any` - to ensure that the conversion resolution error is still correctly reported.

**Comment by JaroslavTulach** [comment]
> Maybe worth adding a few test cases for it?

There were fifteen failing tests in `test/Tests` project when I changed the implementation of `Any.==` - I had to fix them all. E.g. while no test was added, the code is certainly tested.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConversion.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public Function resolveFor(Type into, Type from) {
         current = current.getSupertype();
       }
     }
-    return null;
+    return scope.lookupConversionDefinition(ctx.getBuiltins().any(), into);
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConversion.java:55]: Isn't this going to eliminate the need for all those `....Any.Any` imports introduced by @Akirathan ?

**Comment by radeusgd** [comment]
> > Maybe worth adding a few test cases for it?
> 
> There were fifteen failing tests in `test/Tests` project when I changed the implementation of `Any.==` - I had to fix them all. E.g. while no test was added, the code is certainly tested.

Sure, but these explicit tests would also serve as documentation on what is the behaviour we currently expect.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConversion.java (hunk 1 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConversion.java:55]: Nope. `Comparable.from (_:Any) = Default_Comparator` is defined in `Ordering.enso`. So the minimal import for, say, this snippet `5 + 3 == 8` should be `from Standard.Base.Data.Ordering import all`.

**Review by hubertp — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConversion.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConversion.java:55]: I was wondering: _"how will I find these `....Any.Any` imports?"_

> So the minimal import ... should be `from Standard.Base.Data.Ordering import all`.

Now I no longer need to solve that question....

**Comment by JaroslavTulach** [comment]
> Maybe worth adding a few test cases for it?

There is a lot of tests in `Conversion_Spec.enso` including those that you propose:

> Have a conversion both on type `Foo` and `Any`: verify that for an instance of type `Foo` the more specific one (`Foo`, not `Any`) is selected.

`Not_Foo` type & its tests verify this.

>  Have a conversion on `Any` but not `Foo` - check that `Any` is selected - essentially what this PR does and is implicitly tested by the `==` not breaking, but I think it's good to have an explicit test for it.

That's _"should be able to convert Any"_ test, I'd say.

>  Have no conversion defined on neither `Foo` nor `Any` - to ensure that the conversion resolution error is still correctly reported.

There are two _"should fail graciously..."_ tests that seem to cover this behavior.

At the end I am just adding two tests for foreign Java object and a type: [40cc46b](https://github.com/enso-org/enso/pull/6184/commits/40cc46b93c737742818de69c586cc5ad2e8096c6)

**Comment by JaroslavTulach** [comment]
Running [benchmarks for this PR](https://github.com/enso-org/enso/actions/runs/4603946284/jobs/8134424624). Manual comparison with `develop` branch shows no slowdowns, and indicates few speedups:

obrazek

**Review by radeusgd — APPROVED** [review]

**Pull request merged**